File: bigclaw-go/scripts/migration/live_shadow_scorecard.py
#!/usr/bin/env python3
import argparse
import json
import pathlib
from datetime import datetime, timezone


TIMELINE_DRIFT_TOLERANCE_SECONDS = 0.25
EVIDENCE_FRESHNESS_SLO_HOURS = 168


def utc_iso(moment=None):
    value = moment or datetime.now(timezone.utc)
    return value.isoformat().replace('+00:00', 'Z')


def load_json(path):
    return json.loads(path.read_text(encoding='utf-8'))


def resolve_repo_path(repo_root, path):
    candidate = pathlib.Path(path)
    if candidate.is_absolute():
        return candidate
    return repo_root / candidate


def parse_time(value):
    return datetime.fromisoformat(str(value).replace('Z', '+00:00'))


def collect_event_timestamps(events):
    timestamps = []
    for event in events:
        timestamp = event.get('timestamp')
        if not timestamp:
            continue
        timestamps.append(parse_time(timestamp))
    return timestamps


def latest_report_timestamp(report):
    timestamps = []
    if 'results' in report:
        for item in report['results']:
            timestamps.extend(collect_event_timestamps(item.get('primary', {}).get('events', [])))
            timestamps.extend(collect_event_timestamps(item.get('shadow', {}).get('events', [])))
    else:
        timestamps.extend(collect_event_timestamps(report.get('primary', {}).get('events', [])))
        timestamps.extend(collect_event_timestamps(report.get('shadow', {}).get('events', [])))
    return max(timestamps) if timestamps else None


def build_freshness_entry(name, path, report, generated_at):
    latest_timestamp = latest_report_timestamp(report)
    age_hours = None
    status = 'missing-timestamps'
    if latest_timestamp is not None:
        age_hours = round((generated_at - latest_timestamp).total_seconds() / 3600, 2)
        status = 'fresh' if age_hours <= EVIDENCE_FRESHNESS_SLO_HOURS else 'stale'
    return {
        'name': name,
        'report_path': path,
        'latest_evidence_timestamp': utc_iso(latest_timestamp) if latest_timestamp else None,
        'age_hours': age_hours,
        'freshness_slo_hours': EVIDENCE_FRESHNESS_SLO_HOURS,
        'status': status,
    }


def classify_parity(diff):
    reasons = []
    if not diff.get('state_equal'):
        reasons.append('terminal-state-mismatch')
    if not diff.get('event_types_equal'):
        reasons.append('event-sequence-mismatch')
    event_count_delta = int(diff.get('event_count_delta', 0))
    if event_count_delta != 0:
        reasons.append('event-count-drift')
    timeline_delta = round(
        abs(float(diff.get('primary_timeline_seconds', 0)) - float(diff.get('shadow_timeline_seconds', 0))),
        6,
    )
    if timeline_delta > TIMELINE_DRIFT_TOLERANCE_SECONDS:
        reasons.append('timeline-drift')
    return {
        'status': 'parity-ok' if not reasons else 'drift-detected',
        'timeline_delta_seconds': timeline_delta,
        'timeline_drift_tolerance_seconds': TIMELINE_DRIFT_TOLERANCE_SECONDS,
        'reasons': reasons,
    }


def build_compare_entry(report):
    parity = classify_parity(report.get('diff', {}))
    return {
        'entry_type': 'single-compare',
        'label': 'single fixture compare',
        'trace_id': report.get('trace_id'),
        'source_file': None,
        'source_kind': 'fixture',
        'parity': parity,
        'primary_task_id': report.get('primary', {}).get('task_id'),
        'shadow_task_id': report.get('shadow', {}).get('task_id'),
    }


def build_matrix_entries(report):
    entries = []
    for item in report.get('results', []):
        parity = classify_parity(item.get('diff', {}))
        entries.append(
            {
                'entry_type': 'matrix-row',
                'label': item.get('source_file'),
                'trace_id': item.get('trace_id'),
                'source_file': item.get('source_file'),
                'source_kind': item.get('source_kind'),
                'task_shape': item.get('task_shape'),
                'corpus_slice': item.get('corpus_slice'),
                'parity': parity,
                'primary_task_id': item.get('primary', {}).get('task_id'),
                'shadow_task_id': item.get('shadow', {}).get('task_id'),
            }
        )
    return entries


def check(name, passed, detail):
    return {'name': name, 'passed': passed, 'detail': detail}


def build_report(
    shadow_compare_report_path='bigclaw-go/docs/reports/shadow-compare-report.json',
    shadow_matrix_report_path='bigclaw-go/docs/reports/shadow-matrix-report.json',
):
    repo_root = pathlib.Path(__file__).resolve().parents[3]
    compare_path = resolve_repo_path(repo_root, shadow_compare_report_path)
    matrix_path = resolve_repo_path(repo_root, shadow_matrix_report_path)
    compare_report = load_json(compare_path)
    matrix_report = load_json(matrix_path)
    generated_at = datetime.now(timezone.utc)

    parity_entries = [build_compare_entry(compare_report)] + build_matrix_entries(matrix_report)
    parity_ok_count = sum(1 for item in parity_entries if item['parity']['status'] == 'parity-ok')
    drift_detected_count = len(parity_entries) - parity_ok_count

    freshness = [
        build_freshness_entry('shadow-compare-report', shadow_compare_report_path, compare_report, generated_at),
        build_freshness_entry('shadow-matrix-report', shadow_matrix_report_path, matrix_report, generated_at),
    ]
    stale_inputs = [item for item in freshness if item['status'] != 'fresh']
    latest_evidence_timestamp = max(
        (item['latest_evidence_timestamp'] for item in freshness if item['latest_evidence_timestamp']),
        default=None,
    )

    matrix_corpus_coverage = matrix_report.get('corpus_coverage', {})
    cutover_checkpoints = [
        check(
            'single_compare_matches_terminal_state_and_event_sequence',
            compare_report.get('diff', {}).get('state_equal') and compare_report.get('diff', {}).get('event_types_equal'),
            f"trace_id={compare_report.get('trace_id')}",
        ),
        check(
            'matrix_reports_no_state_or_event_sequence_mismatches',
            matrix_report.get('mismatched', 0) == 0,
            f"matched={matrix_report.get('matched')} mismatched={matrix_report.get('mismatched')}",
        ),
        check(
            'scorecard_detects_no_parity_drift',
            drift_detected_count == 0,
            f"parity_ok={parity_ok_count} drift_detected={drift_detected_count}",
        ),
        check(
            'checked_in_evidence_is_fresh_enough_for_review',
            not stale_inputs,
            f"freshness_statuses={[item['status'] for item in freshness]}",
        ),
        check(
            'matrix_includes_corpus_coverage_overlay',
            bool(matrix_corpus_coverage),
            f"corpus_slice_count={matrix_corpus_coverage.get('corpus_slice_count', 0)}",
        ),
    ]

    limitations = [
        'repo-native only: this scorecard summarizes checked-in shadow artifacts rather than an always-on production traffic mirror',
        'parity drift is measured from fixture-backed compare/matrix runs and optional corpus slices, not mirrored tenant traffic',
        'freshness is derived from the latest artifact event timestamps and should be treated as evidence recency, not live service health',
    ]
    future_work = [
        'replace offline fixture submission with a real ingress mirror or tenant-scoped shadow routing control before treating this as cutover-proof traffic parity',
        'promote parity drift review from checked-in artifacts into a continuously refreshed operational surface',
        'pair this scorecard with rollback automation only after tenant-scoped rollback guardrails exist',
    ]

    return {
        'generated_at': utc_iso(generated_at),
        'ticket': 'BIG-PAR-092',
        'title': 'Live shadow mirror parity drift scorecard',
        'status': 'repo-native-live-shadow-scorecard',
        'evidence_inputs': {
            'shadow_compare_report_path': shadow_compare_report_path,
            'shadow_matrix_report_path': shadow_matrix_report_path,
            'generator_script': 'bigclaw-go/scripts/migration/live_shadow_scorecard',
        },
        'summary': {
            'total_evidence_runs': len(parity_entries),
            'parity_ok_count': parity_ok_count,
            'drift_detected_count': drift_detected_count,
            'matrix_total': matrix_report.get('total', 0),
            'matrix_matched': matrix_report.get('matched', 0),
            'matrix_mismatched': matrix_report.get('mismatched', 0),
            'corpus_coverage_present': bool(matrix_corpus_coverage),
            'corpus_uncovered_slice_count': matrix_corpus_coverage.get('uncovered_corpus_slice_count'),
            'latest_evidence_timestamp': latest_evidence_timestamp,
            'fresh_inputs': sum(1 for item in freshness if item['status'] == 'fresh'),
            'stale_inputs': len(stale_inputs),
        },
        'freshness': freshness,
        'parity_entries': parity_entries,
        'cutover_checkpoints': cutover_checkpoints,
        'limitations': limitations,
        'future_work': future_work,
    }


def main():
    parser = argparse.ArgumentParser(description='Generate a repo-native live shadow mirror parity drift scorecard')
    parser.add_argument('--shadow-compare-report', default='bigclaw-go/docs/reports/shadow-compare-report.json')
    parser.add_argument('--shadow-matrix-report', default='bigclaw-go/docs/reports/shadow-matrix-report.json')
    parser.add_argument('--output', default='bigclaw-go/docs/reports/live-shadow-mirror-scorecard.json')
    parser.add_argument('--pretty', action='store_true')
    args = parser.parse_args()

    repo_root = pathlib.Path(__file__).resolve().parents[3]
    report = build_report(
        shadow_compare_report_path=args.shadow_compare_report,
        shadow_matrix_report_path=args.shadow_matrix_report,
    )
    output_path = resolve_repo_path(repo_root, args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(json.dumps(report, indent=2) + '\n', encoding='utf-8')
    if args.pretty:
        print(json.dumps(report, indent=2))


if __name__ == '__main__':
    main()

File: bigclaw-go/internal/migration/live_shadow_scorecard.go
package migration

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

const (
	TimelineDriftToleranceSeconds = 0.25
	EvidenceFreshnessSLOHours     = 168
)

type Event struct {
	Timestamp string `json:"timestamp"`
}

type TaskRun struct {
	TaskID string  `json:"task_id"`
	Events []Event `json:"events"`
}

type Diff struct {
	StateEqual             bool    `json:"state_equal"`
	EventTypesEqual        bool    `json:"event_types_equal"`
	EventCountDelta        int     `json:"event_count_delta"`
	PrimaryTimelineSeconds float64 `json:"primary_timeline_seconds"`
	ShadowTimelineSeconds  float64 `json:"shadow_timeline_seconds"`
}

type CompareReport struct {
	TraceID string  `json:"trace_id"`
	Primary TaskRun `json:"primary"`
	Shadow  TaskRun `json:"shadow"`
	Diff    Diff    `json:"diff"`
}

type MatrixResult struct {
	TraceID     string  `json:"trace_id"`
	Primary     TaskRun `json:"primary"`
	Shadow      TaskRun `json:"shadow"`
	Diff        Diff    `json:"diff"`
	SourceFile  string  `json:"source_file"`
	SourceKind  string  `json:"source_kind"`
	TaskShape   string  `json:"task_shape"`
	CorpusSlice string  `json:"corpus_slice"`
}

type CorpusCoverage struct {
	CorpusSliceCount          int `json:"corpus_slice_count"`
	UncoveredCorpusSliceCount int `json:"uncovered_corpus_slice_count"`
}

type MatrixReport struct {
	Total          int            `json:"total"`
	Matched        int            `json:"matched"`
	Mismatched     int            `json:"mismatched"`
	Results        []MatrixResult `json:"results"`
	CorpusCoverage CorpusCoverage `json:"corpus_coverage"`
}

type Scorecard struct {
	GeneratedAt    string                 `json:"generated_at"`
	Ticket         string                 `json:"ticket"`
	Title          string                 `json:"title"`
	Status         string                 `json:"status"`
	EvidenceInputs scorecardEvidenceInput `json:"evidence_inputs"`
	Summary        scorecardSummary       `json:"summary"`
	Freshness      []freshnessEntry       `json:"freshness"`
	ParityEntries  []parityEntry          `json:"parity_entries"`
	CutoverChecks  []cutoverCheckpoint    `json:"cutover_checkpoints"`
	Limitations    []string               `json:"limitations"`
	FutureWork     []string               `json:"future_work"`
}

type scorecardEvidenceInput struct {
	ShadowCompareReportPath string `json:"shadow_compare_report_path"`
	ShadowMatrixReportPath  string `json:"shadow_matrix_report_path"`
	GeneratorScript         string `json:"generator_script"`
}

type scorecardSummary struct {
	TotalEvidenceRuns         int    `json:"total_evidence_runs"`
	ParityOKCount             int    `json:"parity_ok_count"`
	DriftDetectedCount        int    `json:"drift_detected_count"`
	MatrixTotal               int    `json:"matrix_total"`
	MatrixMatched             int    `json:"matrix_matched"`
	MatrixMismatched          int    `json:"matrix_mismatched"`
	CorpusCoveragePresent     bool   `json:"corpus_coverage_present"`
	CorpusUncoveredSliceCount int    `json:"corpus_uncovered_slice_count,omitempty"`
	LatestEvidenceTimestamp   string `json:"latest_evidence_timestamp,omitempty"`
	FreshInputs               int    `json:"fresh_inputs"`
	StaleInputs               int    `json:"stale_inputs"`
}

type freshnessEntry struct {
	Name                    string   `json:"name"`
	ReportPath              string   `json:"report_path"`
	LatestEvidenceTimestamp *string  `json:"latest_evidence_timestamp"`
	AgeHours                *float64 `json:"age_hours"`
	FreshnessSLOHours       int      `json:"freshness_slo_hours"`
	Status                  string   `json:"status"`
}

type parityEntry struct {
	EntryType     string       `json:"entry_type"`
	Label         string       `json:"label"`
	TraceID       string       `json:"trace_id"`
	SourceFile    *string      `json:"source_file"`
	SourceKind    *string      `json:"source_kind"`
	TaskShape     *string      `json:"task_shape,omitempty"`
	CorpusSlice   *string      `json:"corpus_slice,omitempty"`
	Parity        parityStatus `json:"parity"`
	PrimaryTaskID string       `json:"primary_task_id"`
	ShadowTaskID  string       `json:"shadow_task_id"`
}

type parityStatus struct {
	Status                        string   `json:"status"`
	TimelineDeltaSeconds          float64  `json:"timeline_delta_seconds"`
	TimelineDriftToleranceSeconds float64  `json:"timeline_drift_tolerance_seconds"`
	Reasons                       []string `json:"reasons"`
}

type cutoverCheckpoint struct {
	Name   string `json:"name"`
	Passed bool   `json:"passed"`
	Detail string `json:"detail"`
}

func BuildLiveShadowScorecard(repoRoot, comparePath, matrixPath string, now time.Time) (Scorecard, error) {
	compare := CompareReport{}
	if err := readJSON(filepath.Join(repoRoot, comparePath), &compare); err != nil {
		return Scorecard{}, err
	}
	matrix := MatrixReport{}
	if err := readJSON(filepath.Join(repoRoot, matrixPath), &matrix); err != nil {
		return Scorecard{}, err
	}

	generatedAt := now.UTC()
	parityEntries := make([]parityEntry, 0, len(matrix.Results)+1)
	parityEntries = append(parityEntries, parityEntry{
		EntryType:     "single-compare",
		Label:         "single fixture compare",
		TraceID:       compare.TraceID,
		SourceFile:    nil,
		SourceKind:    stringPtr("fixture"),
		Parity:        classifyParity(compare.Diff),
		PrimaryTaskID: compare.Primary.TaskID,
		ShadowTaskID:  compare.Shadow.TaskID,
	})
	for _, item := range matrix.Results {
		parityEntries = append(parityEntries, parityEntry{
			EntryType:     "matrix-row",
			Label:         item.SourceFile,
			TraceID:       item.TraceID,
			SourceFile:    nilIfEmpty(item.SourceFile),
			SourceKind:    nilIfEmpty(item.SourceKind),
			TaskShape:     nilIfEmpty(item.TaskShape),
			CorpusSlice:   nilIfEmpty(item.CorpusSlice),
			Parity:        classifyParity(item.Diff),
			PrimaryTaskID: item.Primary.TaskID,
			ShadowTaskID:  item.Shadow.TaskID,
		})
	}

	parityOKCount := 0
	for _, item := range parityEntries {
		if item.Parity.Status == "parity-ok" {
			parityOKCount++
		}
	}

	freshness := []freshnessEntry{
		buildFreshnessEntry("shadow-compare-report", comparePath, latestCompareTimestamp(compare), generatedAt),
		buildFreshnessEntry("shadow-matrix-report", matrixPath, latestMatrixTimestamp(matrix), generatedAt),
	}
	staleInputs := 0
	var latestEvidence *time.Time
	for _, item := range freshness {
		if item.Status != "fresh" {
			staleInputs++
		}
		if item.LatestEvidenceTimestamp == nil {
			continue
		}
		parsed, err := time.Parse(time.RFC3339Nano, *item.LatestEvidenceTimestamp)
		if err != nil {
			continue
		}
		if latestEvidence == nil || parsed.After(*latestEvidence) {
			value := parsed
			latestEvidence = &value
		}
	}

	scorecard := Scorecard{
		GeneratedAt: formatTimestamp(generatedAt),
		Ticket:      "BIG-PAR-092",
		Title:       "Live shadow mirror parity drift scorecard",
		Status:      "repo-native-live-shadow-scorecard",
		EvidenceInputs: scorecardEvidenceInput{
			ShadowCompareReportPath: comparePath,
			ShadowMatrixReportPath:  matrixPath,
			GeneratorScript:         "go run ./cmd/bigclawctl migration live-shadow-scorecard",
		},
		Summary: scorecardSummary{
			TotalEvidenceRuns:         len(parityEntries),
			ParityOKCount:             parityOKCount,
			DriftDetectedCount:        len(parityEntries) - parityOKCount,
			MatrixTotal:               matrix.Total,
			MatrixMatched:             matrix.Matched,
			MatrixMismatched:          matrix.Mismatched,
			CorpusCoveragePresent:     matrix.CorpusCoverage != (CorpusCoverage{}),
			CorpusUncoveredSliceCount: matrix.CorpusCoverage.UncoveredCorpusSliceCount,
			LatestEvidenceTimestamp:   formatTimePointer(latestEvidence),
			FreshInputs:               len(freshness) - staleInputs,
			StaleInputs:               staleInputs,
		},
		Freshness:     freshness,
		ParityEntries: parityEntries,
		CutoverChecks: []cutoverCheckpoint{
			{
				Name:   "single_compare_matches_terminal_state_and_event_sequence",
				Passed: compare.Diff.StateEqual && compare.Diff.EventTypesEqual,
				Detail: fmt.Sprintf("trace_id=%s", compare.TraceID),
			},
			{
				Name:   "matrix_reports_no_state_or_event_sequence_mismatches",
				Passed: matrix.Mismatched == 0,
				Detail: fmt.Sprintf("matched=%d mismatched=%d", matrix.Matched, matrix.Mismatched),
			},
			{
				Name:   "scorecard_detects_no_parity_drift",
				Passed: len(parityEntries)-parityOKCount == 0,
				Detail: fmt.Sprintf("parity_ok=%d drift_detected=%d", parityOKCount, len(parityEntries)-parityOKCount),
			},
			{
				Name:   "checked_in_evidence_is_fresh_enough_for_review",
				Passed: staleInputs == 0,
				Detail: fmt.Sprintf("freshness_statuses=%v", freshnessStatuses(freshness)),
			},
			{
				Name:   "matrix_includes_corpus_coverage_overlay",
				Passed: matrix.CorpusCoverage != (CorpusCoverage{}),
				Detail: fmt.Sprintf("corpus_slice_count=%d", matrix.CorpusCoverage.CorpusSliceCount),
			},
		},
		Limitations: []string{
			"repo-native only: this scorecard summarizes checked-in shadow artifacts rather than an always-on production traffic mirror",
			"parity drift is measured from fixture-backed compare/matrix runs and optional corpus slices, not mirrored tenant traffic",
			"freshness is derived from the latest artifact event timestamps and should be treated as evidence recency, not live service health",
		},
		FutureWork: []string{
			"replace offline fixture submission with a real ingress mirror or tenant-scoped shadow routing control before treating this as cutover-proof traffic parity",
			"promote parity drift review from checked-in artifacts into a continuously refreshed operational surface",
			"pair this scorecard with rollback automation only after tenant-scoped rollback guardrails exist",
		},
	}
	return scorecard, nil
}

func WriteLiveShadowScorecard(path string, report Scorecard) error {
	body, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return err
	}
	body = append(body, '\n')
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, body, 0o644)
}

func readJSON(path string, target any) error {
	body, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("read %s: %w", path, err)
	}
	if err := json.Unmarshal(body, target); err != nil {
		return fmt.Errorf("decode %s: %w", path, err)
	}
	return nil
}

func latestCompareTimestamp(report CompareReport) *time.Time {
	return latestRunTimestamp([]TaskRun{report.Primary, report.Shadow})
}

func latestMatrixTimestamp(report MatrixReport) *time.Time {
	runs := make([]TaskRun, 0, len(report.Results)*2)
	for _, item := range report.Results {
		runs = append(runs, item.Primary, item.Shadow)
	}
	return latestRunTimestamp(runs)
}

func latestRunTimestamp(runs []TaskRun) *time.Time {
	var latest *time.Time
	for _, run := range runs {
		for _, event := range run.Events {
			if event.Timestamp == "" {
				continue
			}
			parsed, err := time.Parse(time.RFC3339Nano, event.Timestamp)
			if err != nil {
				continue
			}
			if latest == nil || parsed.After(*latest) {
				value := parsed
				latest = &value
			}
		}
	}
	return latest
}

func buildFreshnessEntry(name, reportPath string, latest *time.Time, generatedAt time.Time) freshnessEntry {
	entry := freshnessEntry{
		Name:              name,
		ReportPath:        reportPath,
		FreshnessSLOHours: EvidenceFreshnessSLOHours,
		Status:            "missing-timestamps",
	}
	if latest == nil {
		return entry
	}
	ageHours := roundTo((generatedAt.Sub(*latest).Hours()), 2)
	timestamp := formatTimestamp(*latest)
	entry.LatestEvidenceTimestamp = &timestamp
	entry.AgeHours = &ageHours
	if ageHours <= EvidenceFreshnessSLOHours {
		entry.Status = "fresh"
	} else {
		entry.Status = "stale"
	}
	return entry
}

func classifyParity(diff Diff) parityStatus {
	reasons := make([]string, 0, 4)
	if !diff.StateEqual {
		reasons = append(reasons, "terminal-state-mismatch")
	}
	if !diff.EventTypesEqual {
		reasons = append(reasons, "event-sequence-mismatch")
	}
	if diff.EventCountDelta != 0 {
		reasons = append(reasons, "event-count-drift")
	}
	timelineDelta := roundTo(abs(diff.PrimaryTimelineSeconds-diff.ShadowTimelineSeconds), 6)
	if timelineDelta > TimelineDriftToleranceSeconds {
		reasons = append(reasons, "timeline-drift")
	}
	status := "parity-ok"
	if len(reasons) > 0 {
		status = "drift-detected"
	}
	return parityStatus{
		Status:                        status,
		TimelineDeltaSeconds:          timelineDelta,
		TimelineDriftToleranceSeconds: TimelineDriftToleranceSeconds,
		Reasons:                       reasons,
	}
}

func freshnessStatuses(entries []freshnessEntry) []string {
	values := make([]string, 0, len(entries))
	for _, item := range entries {
		values = append(values, item.Status)
	}
	return values
}

func stringPtr(value string) *string {
	return &value
}

func nilIfEmpty(value string) *string {
	if value == "" {
		return nil
	}
	return &value
}

func formatTimestamp(value time.Time) string {
	return value.Format(time.RFC3339Nano)
}

func formatTimePointer(value *time.Time) string {
	if value == nil {
		return ""
	}
	return formatTimestamp(*value)
}

func abs(value float64) float64 {
	if value < 0 {
		return -value
	}
	return value
}

func roundTo(value float64, digits int) float64 {
	scale := 1.0
	for i := 0; i < digits; i++ {
		scale *= 10
	}
	if value >= 0 {
		return float64(int(value*scale+0.5)) / scale
	}
	return float64(int(value*scale-0.5)) / scale
}
